server/src/services/media.service.ts

```diff
@@ -204,33 +225,59 @@ export class MediaService extends BaseService {
+     */
+    let convertedPath: string | undefined;
+    let shouldConvertFromRaw = false;
+    if (imageIsRaw) {
+      let useExtracted = false;
+      // extracted image from RAW is always in JPEG format, as implied from the `jpgFromRaw` tag name
+      convertedPath = StorageCore.getImagePath(asset, AssetPathType.CONVERTED, ImageFormat.JPEG);
+      if (image.extractEmbedded) {
+        // try extracting embedded preview from RAW
+        const didExtract = await this.mediaRepository.extract(asset.originalPath, convertedPath);
+        useExtracted = didExtract && (await this.shouldUseExtractedImage(convertedPath, image.preview.size));
+      }
 
-    try {
-      const useExtracted = didExtract && (await this.shouldUseExtractedImage(extractedPath, image.preview.size));
-      const inputPath = useExtracted ? extractedPath : asset.originalPath;
-      const colorspace = this.isSRGB(asset) ? Colorspace.SRGB : image.colorspace;
-      const processInvalidImages = process.env.IMMICH_PROCESS_INVALID_IMAGES === 'true';
-
-      const orientation = useExtracted && asset.exifInfo?.orientation ? Number(asset.exifInfo.orientation) : undefined;
-      const decodeOptions = { colorspace, processInvalidImages, size: image.preview.size, orientation };
-      const { data, info } = await this.mediaRepository.decodeImage(inputPath, decodeOptions);
-
-      const options = { colorspace, processInvalidImages, raw: info };
-      const outputs = await Promise.all([
-        this.mediaRepository.generateThumbnail(data, { ...image.thumbnail, ...options }, thumbnailPath),
-        this.mediaRepository.generateThumbnail(data, { ...image.preview, ...options }, previewPath),
-        this.mediaRepository.generateThumbhash(data, options),
-      ]);
-
-      return { previewPath, thumbnailPath, thumbhash: outputs[2] };
-    } finally {
-      if (didExtract) {
-        await this.storageRepository.unlink(extractedPath);
+      if (useExtracted) {
+        fullsizePath = convertedPath;
+        // proper orientation metadata is missing in extracted images, specify separately
+        // TODO: remove this when proper EXIF is included in extracted images
+        decodeOptions.orientation = asset.exifInfo?.orientation ? Number(asset.exifInfo.orientation) : undefined;
+      } else {
+        // did not extract or extracted preview is smaller than target size,
+        // convert a full-sized thumbnail from original instead
+        convertedPath = StorageCore.getImagePath(asset, AssetPathType.CONVERTED, image.preview.format);
+        // unset size to disable resizing
+        decodeOptions.size = undefined;
+        shouldConvertFromRaw = true;
       }
     }
+
+    const { info, data } = await this.mediaRepository.decodeImage(fullsizePath, decodeOptions);
+
+    const thumbnailOptions = { colorspace, processInvalidImages, raw: info };
+    const outputs = await Promise.all([
+      this.mediaRepository.generateThumbhash(data, thumbnailOptions),
```

Comment:
This is indeed way slower when `data` is a full-sized buffer from RAW image, even times out on e2e `should upload and generate a thumbnail for different file types`.
Might be faster to sequentially zoom down the files, instead of delaying everything in a concurrency, which also makes the control flow ugly (note the `shouldConvertFromRaw`)

Reply:
Decoding the RAW image is *very* slow, so it's best to only do it once. This is also better in terms of minimizing disk IO (often the bottleneck during bulk import).

But it's also true that these tasks don't require such a high resolution input. Maybe it can be handled in two phases - use the full size buffer for the converted image and resize the buffer to be preview-sized as the input for the others.

Reply:
Alternatively, you can remove this code path and make a separate PR for it, potentially making it a separate setting. There's no need to block this PR for it.

Reply:
Let's merge this first, and I'll make another MR to optimize the performance for RAW.